File: strands-py/src/strands/vended_tools/handoff_to_user/handoff_to_user.py
"""``handoff_to_user`` tool: pause the agent to ask the user a structured question.

A thin shim over :meth:`~strands.types.tools.ToolContext.interrupt`. The tool
validates its inputs, then raises a single interrupt whose ``reason`` carries a
:class:`~strands.vended_tools.handoff_to_user.types.HandoffQuestion` payload.
The SDK halts the agent; the caller resumes with the user's response, which is
threaded back as the tool result.

No console I/O and no UI. Rendering the question and collecting the answer are
the consumer's job (a chat frontend, a Slack app, a CLI handler, etc.). If nothing
is listening the SDK's default interrupt handling applies.

Example:
    ```python
    from strands import Agent
    from strands.vended_tools import handoff_to_user

    agent = Agent(tools=[handoff_to_user])
    result = agent("Ask me which environment to deploy to.")

    if result.stop_reason == "interrupt":
        interrupt = result.interrupts[0]
        # interrupt.reason = {"question": "...", "options": [...], "allow_free_text": ...}
        responses = [
            {
                "interruptResponse": {
                    "interruptId": interrupt.id,
                    "response": {"answer": "prod", "chose": "prod"},
                }
            }
        ]
        result = agent(responses)
    ```
"""

from __future__ import annotations

from ...tools.decorator import tool
from ...types.tools import ToolContext
from .types import (
    HANDOFF_TO_USER_DESCRIPTION,
    INTERRUPT_NAME,
    MAX_OPTION_LENGTH,
    MAX_OPTIONS_COUNT,
    MAX_QUESTION_LENGTH,
    HandoffAnswer,
    HandoffQuestion,
)


def _validate_and_build_reason(
    question: str,
    options: list[str] | None,
    allow_free_text: bool,
) -> HandoffQuestion:
    """Validate the tool inputs and produce the interrupt payload.

    Args:
        question: The question the agent wants to ask the user.
        options: Optional multiple-choice options.
        allow_free_text: Whether free-text answers are accepted. Ignored when
            ``options`` is provided (consumers decide whether to also permit
            free text alongside the choice).

    Returns:
        A :class:`HandoffQuestion` dict for use as the interrupt's ``reason``.

    Raises:
        ValueError: If any input violates the size limits, or if the option list
            contains a non-string or a duplicate entry.
    """
    if not isinstance(question, str) or not question.strip():
        raise ValueError("question must be a non-empty string")
    if len(question) > MAX_QUESTION_LENGTH:
        raise ValueError(f"question length ({len(question)}) exceeds maximum allowed length ({MAX_QUESTION_LENGTH})")

    normalized_options: list[str] | None = None
    if options is not None:
        if not isinstance(options, list):
            raise ValueError("options must be a list of strings")
        if len(options) == 0:
            raise ValueError("options must contain at least one entry when provided")
        if len(options) > MAX_OPTIONS_COUNT:
            raise ValueError(f"options count ({len(options)}) exceeds maximum allowed count ({MAX_OPTIONS_COUNT})")
        # Compare options on their trimmed value so ``["yes", "yes "]`` is
        # rejected as a duplicate — the whitespace check above already treats
        # such entries as equivalent (both are non-empty after strip).
        seen: set[str] = set()
        for i, opt in enumerate(options):
            if not isinstance(opt, str):
                raise ValueError(f"options[{i}] must be a string")
            stripped = opt.strip()
            if not stripped:
                raise ValueError(f"options[{i}] must be a non-empty string")
            if len(opt) > MAX_OPTION_LENGTH:
                raise ValueError(
                    f"options[{i}] length ({len(opt)}) exceeds maximum allowed length ({MAX_OPTION_LENGTH})"
                )
            if stripped in seen:
                raise ValueError(f"options[{i}] duplicates an earlier entry: {opt!r}")
            seen.add(stripped)
        normalized_options = list(options)

    if normalized_options is None and not allow_free_text:
        raise ValueError("handoff must accept either options or free text; got neither")

    # allow_free_text is meaningful only when there are no options; document that
    # in the payload rather than mutating the caller's intent so consumers can see
    # both fields on the wire.
    return HandoffQuestion(
        question=question,
        options=normalized_options,
        allow_free_text=bool(allow_free_text),
    )


def _coerce_response(response: object) -> HandoffAnswer:
    """Normalize the resume response into a :class:`HandoffAnswer` shape.

    - A bare string becomes ``{"answer": <string>}``.
    - A dict with an ``answer`` key (and optionally ``chose``) is passed through
      after minimal type checks.
    - Anything else raises: the tool result should be well-typed, not opaque.

    Args:
        response: Whatever the caller resumed the interrupt with.

    Returns:
        A :class:`HandoffAnswer` dict.

    Raises:
        ValueError: If the response cannot be coerced into a well-formed answer.
    """
    # The same 4096-char budget bounds the question going out and the answer
    # coming back — one direction is user prompt, the other is user answer, both
    # get plumbed into model context, so a single per-turn text budget is enough.
    if isinstance(response, str):
        if len(response) > MAX_QUESTION_LENGTH:
            raise ValueError(
                f"handoff response 'answer' length ({len(response)}) exceeds maximum allowed length "
                f"({MAX_QUESTION_LENGTH})"
            )
        return {"answer": response}
    if isinstance(response, dict):
        answer = response.get("answer")
        chose = response.get("chose")
        if not isinstance(answer, str):
            raise ValueError("handoff response 'answer' must be a string")
        if len(answer) > MAX_QUESTION_LENGTH:
            raise ValueError(
                f"handoff response 'answer' length ({len(answer)}) exceeds maximum allowed length "
                f"({MAX_QUESTION_LENGTH})"
            )
        result: HandoffAnswer = {"answer": answer}
        if chose is not None:
            if not isinstance(chose, str):
                raise ValueError("handoff response 'chose' must be a string when provided")
            if len(chose) > MAX_OPTION_LENGTH:
                raise ValueError(
                    f"handoff response 'chose' length ({len(chose)}) exceeds maximum allowed length "
                    f"({MAX_OPTION_LENGTH})"
                )
            result["chose"] = chose
        return result
    raise ValueError(
        f"handoff response must be a string or an object with an 'answer' key; got {type(response).__name__}"
    )


@tool(name="handoff_to_user", description=HANDOFF_TO_USER_DESCRIPTION, context="tool_context")
def handoff_to_user(
    question: str,
    tool_context: ToolContext,
    options: list[str] | None = None,
    allow_free_text: bool = True,
) -> HandoffAnswer:
    """Pause the agent and ask the user a structured question.

    Emits an interrupt whose ``reason`` carries the question, optional multiple-choice
    options, and whether a free-text answer is acceptable. On resume, the caller's
    response is normalized to a :class:`HandoffAnswer` and returned as the tool
    result so the model can continue the loop with the user's answer in hand.

    Args:
        question: The question to ask the user. Must be a non-empty string no longer
            than 4096 characters.
        tool_context: Injected by the framework. Not user-facing.
        options: Optional multiple-choice options (up to 20 entries, each up to 256
            characters, all distinct). When omitted, the tool asks a free-text
            question.
        allow_free_text: Whether the consumer should accept a free-text answer.
            Defaults to True. When ``options`` is provided this flag is passed
            through to the consumer as a hint but does not restrict what the
            consumer resumes with.

    Returns:
        The user's response as a :class:`HandoffAnswer` dict.
    """
    reason = _validate_and_build_reason(question, options, allow_free_text)
    response = tool_context.interrupt(INTERRUPT_NAME, reason=reason)
    return _coerce_response(response)

File: strands-ts/src/vended-tools/handoff-to-user/handoff-to-user.ts
import { tool } from '../../tools/tool-factory.js'
import { z } from 'zod'
import type { JSONValue } from '../../types/json.js'
import {
  HANDOFF_TO_USER_DESCRIPTION,
  INTERRUPT_NAME,
  MAX_OPTIONS_COUNT,
  MAX_OPTION_LENGTH,
  MAX_QUESTION_LENGTH,
  type HandoffAnswer,
  type HandoffQuestion,
} from './types.js'

/**
 * Zod schema for the tool's input. The schema is the single source of truth
 * for the validated input shape and its size caps; the exported
 * `HandoffToUserInput` type is derived from it via `z.input` (the pre-parse
 * shape, so `allow_free_text` remains optional in the caller-facing type).
 *
 * Zod runs before the callback, so oversized or malformed inputs never reach
 * the interrupt path.
 */
export const handoffToUserInputSchema = z
  .object({
    question: z
      .string()
      .min(1, 'question must be a non-empty string')
      .max(MAX_QUESTION_LENGTH, `question length exceeds maximum allowed length (${MAX_QUESTION_LENGTH})`)
      .describe('The question to ask the user.'),
    options: z
      .array(
        z
          .string()
          .max(MAX_OPTION_LENGTH, `options entry length exceeds maximum allowed length (${MAX_OPTION_LENGTH})`)
          .refine((o) => o.trim().length > 0, 'options entries must be non-empty')
      )
      .min(1, 'options must contain at least one entry when provided')
      .max(MAX_OPTIONS_COUNT, `options count exceeds maximum allowed count (${MAX_OPTIONS_COUNT})`)
      .optional()
      .describe(`Optional multiple-choice options (up to ${MAX_OPTIONS_COUNT} entries).`),
    allow_free_text: z
      .boolean()
      .optional()
      .describe('Whether the consumer should accept a free-text answer (defaults to true).'),
  })
  .refine((v) => v.options !== undefined || v.allow_free_text !== false, {
    message: 'handoff must accept either options or free text; got neither',
  })

/**
 * Input accepted by the `handoffToUser` tool, derived from the Zod schema so
 * the type and validation cannot drift.
 */
export type HandoffToUserInput = z.input<typeof handoffToUserInputSchema>

/**
 * Validate the request beyond what Zod covers and build the interrupt payload.
 *
 * Zod already checks length caps; we additionally enforce uniqueness of options,
 * which the schema alone cannot express without pushing the validator surface up.
 */
function buildReason(input: z.infer<typeof handoffToUserInputSchema>): HandoffQuestion {
  const { question, options, allow_free_text } = input

  if (question.trim().length === 0) {
    throw new Error('question must be a non-empty string')
  }

  let normalizedOptions: string[] | null = null
  if (options !== undefined) {
    // Compare options on their trimmed value so `["yes", "yes "]` is rejected
    // as a duplicate — the schema-level non-empty refinement already treats
    // such entries as equivalent (both are non-empty after trim).
    const seen = new Set<string>()
    for (const opt of options) {
      const stripped = opt.trim()
      if (seen.has(stripped)) {
        throw new Error(`options duplicates an earlier entry: ${JSON.stringify(opt)}`)
      }
      seen.add(stripped)
    }
    normalizedOptions = [...options]
  }

  return {
    question,
    options: normalizedOptions,
    allow_free_text: allow_free_text ?? true,
  }
}

/**
 * Normalize the resume response into a well-shaped HandoffAnswer.
 *
 * - A bare string becomes an object of the form &#123; answer: string &#125;.
 * - An object with a string 'answer' (and optionally string 'chose') passes through.
 * - Anything else throws: the tool result should be well-typed, not opaque.
 */
function coerceResponse(response: JSONValue): HandoffAnswer {
  // The same 4096-char budget bounds the question going out and the answer
  // coming back — one direction is user prompt, the other is user answer, both
  // get plumbed into model context, so a single per-turn text budget is enough.
  if (typeof response === 'string') {
    if (response.length > MAX_QUESTION_LENGTH) {
      throw new Error(
        `handoff response 'answer' length (${response.length}) exceeds maximum allowed length (${MAX_QUESTION_LENGTH})`
      )
    }
    return { answer: response }
  }
  if (response !== null && typeof response === 'object' && !Array.isArray(response)) {
    const { answer, chose } = response as { answer?: unknown; chose?: unknown }
    if (typeof answer !== 'string') {
      throw new Error("handoff response 'answer' must be a string")
    }
    if (answer.length > MAX_QUESTION_LENGTH) {
      throw new Error(
        `handoff response 'answer' length (${answer.length}) exceeds maximum allowed length (${MAX_QUESTION_LENGTH})`
      )
    }
    const result: HandoffAnswer = { answer }
    if (chose !== undefined) {
      if (typeof chose !== 'string') {
        throw new Error("handoff response 'chose' must be a string when provided")
      }
      if (chose.length > MAX_OPTION_LENGTH) {
        throw new Error(
          `handoff response 'chose' length (${chose.length}) exceeds maximum allowed length (${MAX_OPTION_LENGTH})`
        )
      }
      result.chose = chose
    }
    return result
  }
  throw new Error(
    "handoff response must be a string or an object with an 'answer' key; " +
      `got ${response === null ? 'null' : typeof response}`
  )
}

/**
 * handoffToUser: pause the agent to ask the user a structured question.
 *
 * A thin shim over context.interrupt. Validates the request, then raises
 * a single interrupt whose 'reason' carries a HandoffQuestion. The SDK
 * halts the agent; the caller resumes with the user's response, which is
 * normalized into a HandoffAnswer and returned as the tool result.
 *
 * No console I/O and no UI: rendering the question and collecting the answer
 * belong to the consumer (a chat frontend, Slack, custom CLI intervention).
 * When nothing is listening, the SDK's default interrupt handling applies.
 *
 * @example
 * ```typescript
 * import { Agent } from '@strands-agents/sdk'
 * import { handoffToUser } from '@strands-agents/sdk/vended-tools/handoff-to-user'
 *
 * const agent = new Agent({ tools: [handoffToUser] })
 * const result = await agent.invoke('Ask me which environment to deploy to.')
 *
 * if (result.stopReason === 'interrupt') {
 *   const interrupt = result.interrupts[0]
 *   // interrupt.reason = { question, options, allow_free_text }
 *   const resumed = await agent.invoke([
 *     new InterruptResponseContent({
 *       interruptId: interrupt.id,
 *       response: { answer: 'prod', chose: 'prod' },
 *     }),
 *   ])
 * }
 * ```
 */
export const handoffToUser = tool({
  name: 'handoff_to_user',
  description: HANDOFF_TO_USER_DESCRIPTION,
  inputSchema: handoffToUserInputSchema,
  callback: (input, context): HandoffAnswer => {
    if (!context) {
      throw new Error('Tool context is required for handoffToUser')
    }
    const reason = buildReason(input)
    // context.interrupt throws InterruptError on the first call (halting the
    // agent) and returns the user's response on resume.
    const response = context.interrupt<JSONValue>({
      name: INTERRUPT_NAME,
      reason: reason as unknown as JSONValue,
    })
    return coerceResponse(response)
  },
})
